File: DialedInCoffee/Domain/Recommendation.swift
import Foundation

public struct CauseContribution: Codable, Hashable, Sendable, Identifiable {
    public var id: UUID
    public var cause: Cause
    public var confidence: Double          // 0...1
    public var ruleIDs: [String]           // contributing rule ids

    public init(
        id: UUID = UUID(),
        cause: Cause,
        confidence: Double,
        ruleIDs: [String]
    ) {
        self.id = id
        self.cause = cause
        self.confidence = confidence
        self.ruleIDs = ruleIDs
    }
}

public struct Recommendation: Codable, Hashable, Sendable, Identifiable {
    public var id: UUID
    public var primary: Adjustment?
    public var secondary: Adjustment?
    public var topCause: Cause?
    public var confidence: Double          // 0...1
    public var rationale: String
    public var contributions: [CauseContribution]
    public var alternativeCauses: [CauseContribution]
    public var suggestRevertToLastGood: Bool

    public init(
        id: UUID = UUID(),
        primary: Adjustment? = nil,
        secondary: Adjustment? = nil,
        topCause: Cause? = nil,
        confidence: Double = 0,
        rationale: String = "",
        contributions: [CauseContribution] = [],
        alternativeCauses: [CauseContribution] = [],
        suggestRevertToLastGood: Bool = false
    ) {
        self.id = id
        self.primary = primary
        self.secondary = secondary
        self.topCause = topCause
        self.confidence = confidence
        self.rationale = rationale
        self.contributions = contributions
        self.alternativeCauses = alternativeCauses
        self.suggestRevertToLastGood = suggestRevertToLastGood
    }

    public var adjustments: [Adjustment] {
        [primary, secondary].compactMap { $0 }
    }

    public var hasAnyAdjustment: Bool { primary != nil }
}

File: android/core/src/main/kotlin/com/cawfee/domain/model/Recommendation.kt
package com.cawfee.domain.model

import java.util.UUID

/** One cause's contribution to a recommendation. Ported from Recommendation.swift. */
data class CauseContribution(
    val id: UUID = UUID.randomUUID(),
    val cause: Cause,
    val confidence: Double,          // 0..1
    val ruleIDs: List<String>,
)

/** The full output of the rules engine. Ported from Recommendation.swift. */
data class Recommendation(
    val id: UUID = UUID.randomUUID(),
    val primary: Adjustment? = null,
    val secondary: Adjustment? = null,
    val topCause: Cause? = null,
    val confidence: Double = 0.0,    // 0..1
    val rationale: String = "",
    val contributions: List<CauseContribution> = emptyList(),
    val alternativeCauses: List<CauseContribution> = emptyList(),
    val suggestRevertToLastGood: Boolean = false,
) {
    val adjustments: List<Adjustment> get() = listOfNotNull(primary, secondary)
    val hasAnyAdjustment: Boolean get() = primary != null
}
